core/src/main/java/org/apache/carbondata/core/carbon/datastore/BlockIndexStore.java

```diff
@@ -16,113 +16,142 @@
 import org.apache.carbondata.core.carbon.datastore.block.TableBlockInfo;
+import org.apache.carbondata.core.carbon.datastore.block.TableBlockUniqueIdentifier;
 import org.apache.carbondata.core.carbon.datastore.exception.IndexBuilderException;
-import org.apache.carbondata.core.carbon.metadata.blocklet.DataFileFooter;
 import org.apache.carbondata.core.constants.CarbonCommonConstants;
 import org.apache.carbondata.core.util.CarbonProperties;
-import org.apache.carbondata.core.util.CarbonUtil;
 import org.apache.carbondata.core.util.CarbonUtilException;
 
 /**
- * Singleton Class to handle loading, unloading,clearing,storing of the table
- * blocks
+ * This class is used to load the B-Tree in Executor LRU Cache
  */
-public class BlockIndexStore {
-
-  /**
-   * singleton instance
-   */
-  private static final BlockIndexStore CARBONTABLEBLOCKSINSTANCE = new BlockIndexStore();
-
-  /**
-   * map to hold the table and its list of blocks
-   */
-  private Map<AbsoluteTableIdentifier, Map<BlockInfo, AbstractIndex>> tableBlocksMap;
-
-  /**
-   * map to maintain segment id to block info map, this map will be used to
-   * while removing the block from memory when segment is compacted or deleted
-   */
-  private Map<AbsoluteTableIdentifier, Map<String, List<BlockInfo>>> segmentIdToBlockListMap;
-
-  /**
-   * map of block info to lock object map, while loading the btree this will be filled
-   * and removed after loading the tree for that particular block info, this will be useful
-   * while loading the tree concurrently so only block level lock will be applied another
-   * block can be loaded concurrently
-   */
-  private Map<BlockInfo, Object> blockInfoLock;
+public class BlockIndexStore<K, V> extends AbstractBlockIndexStoreCache<K, V> {
 
   /**
-   * table and its lock object to this will be useful in case of concurrent
-   * query scenario when more than one query comes for same table and in that
-   * case it will ensure that only one query will able to load the blocks
+   * LOGGER instance
    */
-  private Map<AbsoluteTableIdentifier, Object> tableLockMap;
+  private static final LogService LOGGER =
+      LogServiceFactory.getLogService(BlockIndexStore.class.getName());
+  public BlockIndexStore(String carbonStorePath, CarbonLRUCache lruCache) {
+    super(carbonStorePath, lruCache);
+  }
 
   /**
-   * block info to future task mapping
-   * useful when blocklet distribution is enabled and
-   * same block is loaded by multiple thread
+   * The method loads the block meta in B-tree lru cache and returns the block meta.
+   *
+   * @param tableBlockUniqueIdentifier Uniquely identifies the block
+   * @return returns the blocks B-Tree meta
+   * @throws CarbonUtilException
    */
-  private Map<BlockInfo, Future<AbstractIndex>> mapOfBlockInfoToFuture;
+  @Override public AbstractIndex get(TableBlockUniqueIdentifier tableBlockUniqueIdentifier)
+      throws CarbonUtilException {
```

Comment:
Its bound to the Cache interface methods
V get(K key) throws CarbonUtilException;
List getAll(List keys) throws CarbonUtilException;